File: Classes/core/RootKit/Categories/NSNumber+Formatter.m
//
//  NSNumber+Formatter.m
//  TripKit
//
//  Created by Adrian Schönig on 18/01/12.
//  Copyright (c) 2012 SkedGo. All rights reserved.
//

#import "NSNumber+Formatter.h"

#import "SGStylemanager.h"

@implementation NSNumber (Formatter)

- (NSString *)toMoneyString:(NSString *)currencySymbol
{
  double dollars = ceil([self floatValue]);
  if (dollars < 1) {
    return NSLocalizedStringFromTableInBundle(@"Free", @"Shared", [SGStyleManager bundle], "Free as in beer");
  } else {
    if (! currencySymbol) {
      currencySymbol = @"$";
    }
    
    NSNumberFormatter *formatter = [[self class] numberFormatter];
    formatter.numberStyle = NSNumberFormatterCurrencyStyle;
    formatter.currencySymbol = currencySymbol;
    formatter.roundingIncrement = @(1);
    return [formatter stringFromNumber:self];
  }
}

- (NSString *)toCarbonString
{
//TODO: Use NSMassFormatter in iOS 8 - http://nshipster.com/nsformatter/
  
  if ([self floatValue] == 0.0) {
    return NSLocalizedStringFromTableInBundle(@"No CO₂", @"Shared", [SGStyleManager bundle], nil);
  } else {
    NSNumberFormatter *formatter = [[self class] numberFormatter];
    formatter.numberStyle = NSNumberFormatterDecimalStyle;
    formatter.currencySymbol = nil;
    formatter.roundingIncrement = @(0.1);
    return [NSString stringWithFormat:@"%@kg CO₂", [formatter stringFromNumber:self]];
  }
}

#pragma mark - Helpers

+ (NSNumberFormatter *)numberFormatter {
  static NSNumberFormatter *_numberFormatter = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    _numberFormatter = [[NSNumberFormatter alloc] init];
    _numberFormatter.locale = [NSLocale currentLocale];
    _numberFormatter.maximumSignificantDigits = 2;
    _numberFormatter.usesSignificantDigits = YES;
    _numberFormatter.roundingMode = NSNumberFormatterRoundUp;
  });
  
  return _numberFormatter;
}

@end

File: TripKit/Classes/core/RootKit/Categories/NSNumber+Formatter.swift
//
//  NSNumber+Formatter.swift
//  TripKit
//
//  Created by Adrian Schönig on 10.04.20.
//  Copyright © 2020 SkedGo Pty Ltd. All rights reserved.
//

import Foundation

extension NSNumber {
  
  private static let formatter: NumberFormatter = {
    let formatter = NumberFormatter()
    formatter.locale = .autoupdatingCurrent
    formatter.maximumSignificantDigits = 2
    formatter.usesSignificantDigits = true
    formatter.roundingMode = .up
    return formatter
  }()
  
  @objc(toMoneyStringWithCurrencyCode:)
  public func toMoneyString(currencyCode: String) -> String {
    guard ceil(floatValue) >= 1 else {
      return NSLocalizedString("Free", tableName: "Shared", bundle: .tripKit, comment: "Free as in beer")
    }
    
    let formatter = Self.formatter
    formatter.numberStyle = .currency
    formatter.currencyCode = currencyCode
    formatter.roundingIncrement = NSNumber(value: 1)
    return formatter.string(from: self)!
  }
  
  @objc(toCarbonString)
  public func toCarbonString() -> String {
    guard floatValue > 0 else {
      return NSLocalizedString("No CO₂", tableName: "Shared", bundle: .tripKit, comment: "Indicator for no carbon emissions")
    }
    
    let formatter = Self.formatter
    formatter.numberStyle = .decimal
    formatter.currencyCode = nil
    formatter.roundingIncrement = NSNumber(value: 0.1)
    return NSString(format: "%@kg CO₂", formatter.string(from: self)!) as String
  }
  
}
